File: LC/[0329] Longest Increasing Path in a Matrix.cpp
class Solution {
public:
    int m;
    int n;
    int memo[200][200];
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        m = matrix.size();
        n = matrix[0].size();
        int ret = 0;
        for(int ii=0; ii<m; ++ii){
            for(int jj=0; jj<n; ++jj){
                ret = max(ret, dfs(matrix, ii, jj));
            }
        }
        return ret;    
    }

    int dfs(vector<vector<int>> & matrix, int ii, int jj){
        if(memo[ii][jj] != 0) return memo[ii][jj];
        vector<pair<int, int>> dir{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int ret = 1;
        for(int kk=0; kk<4; ++kk){
            int x = ii+dir[kk].first;
            int y = jj+dir[kk].second;
            if(x < 0 || y < 0 || x >= m || y >= n) continue;
            if(matrix[x][y] <= matrix[ii][jj]) continue;
            ret = max(ret, 1+dfs(matrix, x, y));
        }
        memo[ii][jj] = ret;
        return ret;
    }
};

File: LC/[0329] Longest Increasing Path in a Matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        # we will traverse the matrix and dfs for the longest increasing path
        
        memo = dict()
        m=len(matrix)
        n=len(matrix[0])

        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def dfs(ii, jj):
            if (ii, jj) in memo: return memo[(ii, jj)]
            ans =1
            for dx, dy in directions:
                nx, ny = ii+dx, jj+dy
                if nx < 0 or nx >= m or ny < 0 or ny >=n: continue
                if matrix[nx][ny] <= matrix[ii][jj]: continue
                ans = max(ans, 1+dfs(nx, ny))

            memo[(ii, jj)] = ans
            return ans
        
        res=0
        for ii in range(m):
            for jj in range(n):
                res = max(res, dfs(ii, jj))

        return res
        


        